_stbt/control_gpl.py

```diff
@@ -0,0 +1,217 @@
+        "KEY_FASTFORWARD": 73,
+        "KEY_EJECT": 74,
+        "KEY_FORWARD": 75,
+        "KEY_BACKWARD": 76,
+        "KEY_STOP_RECORD": 77,
+        "KEY_PAUSE_RECORD": 78,
+        "KEY_ANGLE": 80,
+        "KEY_SUB_PICTURE": 81,  # <- not in input-event-codes.h
+        "KEY_VOD": 82,
+        "KEY_EPG": 83,
+        "KEY_TIMER_PROGRAMMING": 84,  # <- not in input-event-codes.h
+        "KEY_CONFIG": 85,
+
+        # Not sure what the difference is between KEY_PLAY and KEY_PLAY_FUNCTION
+        # is but none of these _FUNCTION keys are in linux-event-codes.h:
+        "KEY_PLAY_FUNCTION": 96,
+        "KEY_PAUSE_PLAY_FUNCTION": 97,
+        "KEY_RECORD_FUNCTION": 98,
+        "KEY_PAUSE_RECORD_FUNCTION": 99,
+        "KEY_STOP_FUNCTION": 100,
+        "KEY_MUTE_FUNCTION": 101,
+        "KEY_RESTORE_VOLUME_FUNCTION": 102,
+        "KEY_TUNE_FUNCTION": 103,
+        "KEY_SELECT_MEDIA_FUNCTION": 104,
+        "KEY_SELECT_AV_INPUT_FUNCTION": 105,
+        "KEY_SELECT_AUDIO_INPUT_FUNCTION": 106,
+        "KEY_POWER_TOGGLE_FUNCTION": 107,
+        "KEY_POWER_OFF_FUNCTION": 108,
+        "KEY_POWER_ON_FUNCTION": 109,
+
+        # Back to normal keys.  We duplicate these colour buttons because HDMI
+        # doesn't make a distinction:
+        "KEY_F1": 113,
+        "KEY_BLUE": 113,
+        "KEY_F2": 114,
+        "KEY_RED": 114,
+        "KEY_F3": 115,
+        "KEY_GREEN": 115,
+        "KEY_F4": 116,
+        "KEY_YELLOW": 116,
+
+        # And back to normal keys:
+        "KEY_F5": 117,
+        "KEY_DATA": 118,
+    }
+
+    def __init__(self, device, source, destination):
+        import cec
+        if source is None:
+            source = 1
+        if destination is None:
+            destination = 4
+        self.source = int(source, 16)
+        self.destination = int(destination, 16)
+
+        self.cecconfig = cec.libcec_configuration()
+        self.cecconfig.strDeviceName = "stb-tester"
+        self.cecconfig.bActivateSource = 0
+        self.cecconfig.deviceTypes.Add(cec.CEC_DEVICE_TYPE_RECORDING_DEVICE)
+        self.cecconfig.clientVersion = cec.LIBCEC_VERSION_CURRENT
+        self.lib = cec.ICECAdapter.Create(self.cecconfig)
+        debug("libCEC version %s loaded: %s" % (
+            self.lib.VersionToString(self.cecconfig.serverVersion),
+            self.lib.GetLibInfo()))
+
+        if device is None:
+            device = self.detect_adapter()
+            if device is None:
+                raise HdmiCecError("No adapter found")
+        if not self.lib.Open(device):
+            raise HdmiCecError("Failed to open a connection to the CEC adapter")
+        debug("Connection to CEC adapter opened")
+
+    def press(self, key):
+        from .control import UnknownKeyError
+        keycode = self._KEYNAMES.get(key)
+        if keycode is None:
+            raise UnknownKeyError("HdmiCecControl: Unknown key %r" % key)
+        cec_command = "%X%X:44:%02X" % (self.source, self.destination, keycode)
```

Comment:
I don't want to do this for the time being.  The magic CEC numbers are not in themselves interesting.  In the future I think it would be good to match devices by OSD name as this is more likely to be meaningful and stable.

This would involve doing an CEC scan to work out which device should be used.  This wouldn't be difficult and it can be added in a future PR in a backwards compatible way, but I don't have time to spend on this right now.